File: src/main/hook-installer.ts
/**
 * HookInstaller — 读写 ~/.claude/settings.json
 *
 * 安装: 将 Claude Island 的 HTTP hook 写入 settings.json
 * 卸载: 移除 localhost:PORT 相关的 hook 条目, 保留其他配置
 *
 * Hook 事件分两类:
 * - blocking (120s): PreToolUse, PermissionRequest — 需要等待用户审批
 * - notifying (5s): SessionStart, SessionEnd, PostToolUse 等 — 仅通知
 */

import fs from 'node:fs';
import path from 'node:path';
import os from 'node:os';

const SETTINGS_PATH = path.join(os.homedir(), '.claude', 'settings.json');

const HOOK_EVENTS = {
  blocking: ['PreToolUse', 'PermissionRequest'],
  notifying: [
    'UserPromptSubmit',
    'SessionStart', 'SessionEnd', 'PostToolUse',
    'Notification', 'Stop',
  ],
};

/** 安装 Claude Island hooks 到 ~/.claude/settings.json (合并, 不覆盖已有 hooks) */
export function installHooks(port: number = 51515): void {
  let settings: Record<string, any> = {};
  try {
    const raw = fs.readFileSync(SETTINGS_PATH, 'utf-8');
    settings = JSON.parse(raw);
  } catch {
    // 文件不存在或解析失败
  }

  const url = `http://localhost:${port}/hook`;
  const marker = `localhost:${port}`;
  const existingHooks: Record<string, any[]> = settings.hooks || {};

  // 为每个事件合并 hook: 保留其他来源的 hooks, 追加/替换 cc-island 的
  for (const event of HOOK_EVENTS.blocking) {
    existingHooks[event] = mergeHookEntry(existingHooks[event], url, marker, 120);
  }
  for (const event of HOOK_EVENTS.notifying) {
    existingHooks[event] = mergeHookEntry(existingHooks[event], url, marker, 5);
  }

  settings.hooks = existingHooks;

  fs.mkdirSync(path.dirname(SETTINGS_PATH), { recursive: true });
  fs.writeFileSync(SETTINGS_PATH, JSON.stringify(settings, null, 2));
  console.log(`[HookInstaller] Hooks installed to ${SETTINGS_PATH}`);
}

/** 合并单个事件的 hook 条目: 保留非 cc-island 的, 追加 cc-island 的 */
function mergeHookEntry(existing: any[] | undefined, url: string, marker: string, timeout: number): any[] {
  const others = (existing || []).filter((m: any) => {
    const hooks = m.hooks || [];
    return !hooks.some((h: any) => h.url?.includes(marker));
  });
  others.push({ hooks: [{ type: 'http', url, timeout }] });
  return others;
}

/** 移除 Claude Island hooks, 保留其他配置 */
export function removeHooks(port: number = 51515): void {
  try {
    const raw = fs.readFileSync(SETTINGS_PATH, 'utf-8');
    const settings = JSON.parse(raw);

    if (settings.hooks) {
      const marker = `localhost:${port}`;
      for (const [event, matchers] of Object.entries(settings.hooks)) {
        if (Array.isArray(matchers)) {
          settings.hooks[event] = matchers.filter((m: any) => {
            const hooks = m.hooks || [];
            return !hooks.some((h: any) => h.url?.includes(marker));
          });
          if (settings.hooks[event].length === 0) {
            delete settings.hooks[event];
          }
        }
      }
      if (Object.keys(settings.hooks).length === 0) {
        delete settings.hooks;
      }
    }

    fs.writeFileSync(SETTINGS_PATH, JSON.stringify(settings, null, 2));
    console.log('[HookInstaller] Hooks removed');
  } catch {
    // 文件不存在则忽略
  }
}

/** 检查当前是否已安装 hooks (每次读文件, 不缓存) */
export function isInstalled(port: number = 51515): boolean {
  try {
    const raw = fs.readFileSync(SETTINGS_PATH, 'utf-8');
    const settings = JSON.parse(raw);
    return JSON.stringify(settings.hooks || {}).includes(`localhost:${port}`);
  } catch {
    return false;
  }
}

// CLI 入口
if (require.main === module) {
  const cmd = process.argv[2];
  if (cmd === 'install') {
    installHooks();
  } else if (cmd === 'remove') {
    removeHooks();
  } else {
    console.log('Usage: hook-installer [install|remove]');
  }
}

File: apps/cc-island/src-tauri/src/hook_installer.rs
use std::{fs, path::PathBuf};

use serde_json::{json, Value};

pub fn settings_path() -> PathBuf {
    let home = std::env::var("HOME")
        .or_else(|_| std::env::var("USERPROFILE"))
        .unwrap_or_else(|_| "~".into());
    PathBuf::from(home).join(".claude").join("settings.json")
}

const BLOCKING_EVENTS: &[&str] = &["PreToolUse", "PermissionRequest"];
const NOTIFYING_EVENTS: &[&str] = &[
    "UserPromptSubmit",
    "SessionStart",
    "SessionEnd",
    "PostToolUse",
    "Notification",
    "Stop",
];

pub fn install_hooks(port: u16) -> Result<(), String> {
    let path = settings_path();
    let mut settings: Value = if path.exists() {
        let raw = fs::read_to_string(&path).map_err(|e| e.to_string())?;
        serde_json::from_str(&raw).unwrap_or_else(|_| json!({}))
    } else {
        json!({})
    };

    if !settings.is_object() {
        settings = json!({});
    }

    let url = format!("http://localhost:{}/hook", port);
    let marker = format!("localhost:{}", port);

    if settings.get("hooks").is_none() {
        settings["hooks"] = json!({});
    }

    for event in BLOCKING_EVENTS {
        let existing = settings["hooks"]
            .get(*event)
            .and_then(|v| v.as_array())
            .cloned()
            .unwrap_or_default();
        settings["hooks"][*event] =
            Value::Array(merge_hook_entry(&existing, &url, &marker, 120));
    }

    for event in NOTIFYING_EVENTS {
        let existing = settings["hooks"]
            .get(*event)
            .and_then(|v| v.as_array())
            .cloned()
            .unwrap_or_default();
        settings["hooks"][*event] =
            Value::Array(merge_hook_entry(&existing, &url, &marker, 5));
    }

    if let Some(parent) = path.parent() {
        fs::create_dir_all(parent).map_err(|e| e.to_string())?;
    }
    let pretty = serde_json::to_string_pretty(&settings).map_err(|e| e.to_string())?;
    fs::write(&path, pretty).map_err(|e| e.to_string())?;

    eprintln!("[HookInstaller] Hooks installed to {:?}", path);
    Ok(())
}

fn merge_hook_entry(existing: &[Value], url: &str, marker: &str, timeout: u32) -> Vec<Value> {
    let mut others: Vec<Value> = existing
        .iter()
        .filter(|m| {
            let hooks = m.get("hooks").and_then(|v| v.as_array());
            !hooks
                .map(|hooks| {
                    hooks.iter().any(|h| {
                        h.get("url")
                            .and_then(|u| u.as_str())
                            .map(|u| u.contains(marker))
                            .unwrap_or(false)
                    })
                })
                .unwrap_or(false)
        })
        .cloned()
        .collect();

    others.push(json!({
        "hooks": [{ "type": "http", "url": url, "timeout": timeout }]
    }));

    others
}

pub fn remove_hooks(port: u16) -> Result<(), String> {
    let path = settings_path();
    if !path.exists() {
        return Ok(());
    }

    let raw = fs::read_to_string(&path).map_err(|e| e.to_string())?;
    let mut settings: Value = serde_json::from_str(&raw).unwrap_or_else(|_| json!({}));
    let marker = format!("localhost:{}", port);

    if let Some(hooks) = settings.get_mut("hooks").and_then(|v| v.as_object_mut()) {
        let keys: Vec<String> = hooks.keys().cloned().collect();
        for key in keys {
            if let Some(matchers) = hooks.get_mut(&key).and_then(|v| v.as_array_mut()) {
                matchers.retain(|m| {
                    let hook_arr = m.get("hooks").and_then(|v| v.as_array());
                    !hook_arr
                        .map(|hooks| {
                            hooks.iter().any(|h| {
                                h.get("url")
                                    .and_then(|u| u.as_str())
                                    .map(|u| u.contains(&marker))
                                    .unwrap_or(false)
                            })
                        })
                        .unwrap_or(false)
                });
            }
            if hooks
                .get(&key)
                .and_then(|v| v.as_array())
                .map(|a| a.is_empty())
                .unwrap_or(false)
            {
                hooks.remove(&key);
            }
        }
        if hooks.is_empty() {
            settings
                .as_object_mut()
                .map(|obj| obj.remove("hooks"));
        }
    }

    let pretty = serde_json::to_string_pretty(&settings).map_err(|e| e.to_string())?;
    fs::write(&path, pretty).map_err(|e| e.to_string())?;
    eprintln!("[HookInstaller] Hooks removed");
    Ok(())
}

pub fn is_installed(port: u16) -> bool {
    let path = settings_path();
    let marker = format!("localhost:{}", port);
    match fs::read_to_string(path) {
        Ok(raw) => raw.contains(&marker),
        Err(_) => false,
    }
}
